Add support for updatable list of domains with restricted openers

bnham opened the pull request:
#### ce5ed8691c6da66b81833634d7e01d7888fbd90e
<pre>
Add support for updatable list of domains with restricted openers
<a href="https://bugs.webkit.org/show_bug.cgi?id=268185">https://bugs.webkit.org/show_bug.cgi?id=268185</a>
<a href="https://rdar.apple.com/problem/121677395">rdar://problem/121677395</a>

Reviewed by Wenson Hsieh.

This adds support for a list of domains that will receive restricted window opener access. The list
is maintained and updated through WebPrivacy.

I didn&apos;t add a method to override the list for testing purposes yet. I&apos;ll do that in a future patch
which actually uses these interfaces.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForRestrictedOpenerDomainsChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::RestrictedOpenerDomainsController::shared):
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::restrictedOpenerType):
(WebKit::RestrictedOpenerDomainsController::populate):
(WebKit::RestrictedOpenerDomainsController::lookup const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/273693@main">https://commits.webkit.org/273693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6062f4472eb678807783e078d9b1345e5891d371

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36349 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39069 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37576 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12335 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36909 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12921 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32229 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11324 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32444 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40313 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32986 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32806 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11562 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35381 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8243 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -414,6 +426,89 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
 
+static uint64_t approximateContinuousTimeNanoseconds()
+{
+    return clock_gettime_nsec_np(CLOCK_MONOTONIC_RAW_APPROX);
```
I just wanted an approximate clock that incremented while the system was sleeping, which is `CLOCK_MONOTONIC_RAW_APPROX` or `mach_continuous_approximate_time`, but it looks like we don't have wrappers around those in WebKit right now.

whsieh replied:
I see, this seems fine then.

whsieh commented on Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
```diff
@@ -123,7 +123,7 @@ constexpr NSInteger WPResourceTypeStorageAccessUserAgentStringQuirksData = 6;
 typedef void (^WPStorageAccessUserAgentStringQuirksDataCompletionHandler)(WPStorageAccessUserAgentStringQuirksData *, NSError *);
 
-@interface WPStorageAccessUserAgentStringQuirk (Staging_119342418)
+@interface WPResources (Staging_119342418_UAQuirks)
```
😅

(Nice catch)

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -29,6 +29,7 @@
 #import <wtf/Function.h>
 #import <wtf/Ref.h>
 #import <wtf/RetainPtr.h>
+#import <wtf/Seconds.h>
```
Is this import necessary, since we're just using `uint64_t`?

bnham committed:
Add support for updatable list of domains with restricted openers
https://bugs.webkit.org/show_bug.cgi?id=268185
rdar://problem/121677395

Reviewed by Wenson Hsieh.

This adds support for a list of domains that will receive restricted window opener access. The list
is maintained and updated through WebPrivacy.

I didn't add a method to override the list for testing purposes yet. I'll do that in a future patch
which actually uses these interfaces.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForRestrictedOpenerDomainsChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::RestrictedOpenerDomainsController::shared):
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::restrictedOpenerType):
(WebKit::RestrictedOpenerDomainsController::populate):
(WebKit::RestrictedOpenerDomainsController::lookup const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/273693@main
